libbpf-tools/softirqs.c

```diff
@@ -0,0 +1,258 @@
+
+static const struct argp_option opts[] = {
+	{ "distributed", 'd', NULL, 0, "Show distributions as histograms" },
+	{ "timestamp", 'T', NULL, 0, "Include timestamp on output" },
+	{ "nanoseconds", 'N', NULL, 0, "Output in nanoseconds" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'd':
+		env.distributed = true;
+		break;
+	case 'N':
+		env.nanoseconds = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+enum {
+	HI_SOFTIRQ = 0,
+	TIMER_SOFTIRQ = 1,
+	NET_TX_SOFTIRQ = 2,
+	NET_RX_SOFTIRQ = 3,
+	BLOCK_SOFTIRQ = 4,
+	IRQ_POLL_SOFTIRQ = 5,
+	TASKLET_SOFTIRQ = 6,
+	SCHED_SOFTIRQ = 7,
+	HRTIMER_SOFTIRQ = 8,
+	RCU_SOFTIRQ = 9,
+	NR_SOFTIRQS = 10,
+};
+
+static char *vec_names[] = {
+	[HI_SOFTIRQ] = "hi",
+	[TIMER_SOFTIRQ] = "timer",
+	[NET_TX_SOFTIRQ] = "net_tx",
+	[NET_RX_SOFTIRQ] = "net_rx",
+	[BLOCK_SOFTIRQ] = "block",
+	[IRQ_POLL_SOFTIRQ] = "irq_poll",
+	[TASKLET_SOFTIRQ] = "tasklet",
+	[SCHED_SOFTIRQ] = "sched",
+	[HRTIMER_SOFTIRQ] = "hrtimer",
+	[RCU_SOFTIRQ] = "rcu",
+};
+
+static int print_count(struct softirqs_bpf__bss *bss)
+{
+	const char *units = env.nanoseconds ? "nsecs" : "usecs";
+	__u32 vec;
+
+	printf("%-16s %6s%5s\n", "SOFTIRQ", "TOTAL_", units);
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		if (bss->counts[vec] > 0)
+			printf("%-16s %11llu\n", vec_names[vec],
+				bss->counts[vec]);
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++)
+		__atomic_store_n(&bss->counts[vec], 0, __ATOMIC_RELAXED);
+
+	return 0;
+}
+
+static struct hist zero;
+
+static int print_hist(struct softirqs_bpf__bss *bss)
+{
+	const char *units = env.nanoseconds ? "nsecs" : "usecs";
+	struct hist *hist;
+	__u32 vec;
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++) {
+		hist = &bss->hists[vec];
+		if (!memcmp(&zero, hist, sizeof(*hist)))
+			continue;
+		printf("softirq = %s\n", vec_names[vec]);
+		print_log2_hist(hist->slots, MAX_SLOTS, units);
+		printf("\n");
+	}
+
+	for (vec = 0; vec < NR_SOFTIRQS; vec++)
+		__atomic_store(&bss->hists[vec], &zero, __ATOMIC_RELAXED);
```

Comment:
how is this supposed to work? atomics are only available for a single word integers, not arbitrary structs. I guess for histograms you'll have to use map after all :(

Reply:
or we can just leave with races and zero out non-atomically with memset, which is probably ok as well

Reply:
> how is this supposed to work? atomics are only available for a single word integers, not arbitrary structs. I guess for histograms you'll have to use map after all :(

Oh I think I misunderstood the meaning of "generic version"

> Built-in Function: void __atomic_store (type *ptr, type *val, int memorder)
> This is the generic version of an atomic store. It stores the value of *val into *ptr.

Reply:
> how is this supposed to work? 

Um, I am a bit confused about this at the moment,  I do agree 

> atomics are only available for a single word integers, not arbitrary structs. 

But from  https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html I saw

> Built-in Function: void __atomic_store (type *ptr, type *val, int memorder)
> This is the generic version of an atomic store. It stores the value of *val into *ptr.

And from the test results, it does modify the entire structure （ `softirq -d 1` # clear  once a second）

```
softirq = timer
     usecs               : count    distribution
         0 -> 1          : 600      |****************************************|
         2 -> 3          : 39       |**                                      |
         4 -> 7          : 53       |***                                     |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 13       |                                        |
        32 -> 63         : 3        |                                        |
        64 -> 127        : 2        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 679      |****************************************|
         2 -> 3          : 30       |*                                       |
         4 -> 7          : 50       |**                                      |
         8 -> 15         : 5        |                                        |
        16 -> 31         : 5        |                                        |
        32 -> 63         : 1        |                                        |
        64 -> 127        : 3        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 483      |****************************************|
         2 -> 3          : 30       |**                                      |
         4 -> 7          : 46       |***                                     |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 4        |                                        |
        32 -> 63         : 3        |                                        |
        64 -> 127        : 3        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 426      |****************************************|
         2 -> 3          : 22       |**                                      |
         4 -> 7          : 40       |***                                     |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 2        |                                        |
        32 -> 63         : 4        |                                        |
        64 -> 127        : 1        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 565      |****************************************|
         2 -> 3          : 41       |**                                      |
         4 -> 7          : 44       |***                                     |
         8 -> 15         : 7        |                                        |
        16 -> 31         : 9        |                                        |
        32 -> 63         : 4        |                                        |
        64 -> 127        : 1        |                                        |

softirq = timer
     usecs               : count    distribution
         0 -> 1          : 391      |****************************************|
         2 -> 3          : 22       |**                                      |
         4 -> 7          : 41       |****                                    |
         8 -> 15         : 4        |                                        |
        16 -> 31         : 1        |                                        |
        32 -> 63         : 4        |                                        |
        64 -> 127        : 1        |                                        |
```

I will determine the reason.

Reply:
> how is this supposed to work? atomics are only available for a single word integers, not arbitrary structs. I guess for histograms you'll have to use map after all :(

From
https://sourcegraph.com/github.com/gcc-mirror/gcc@e1336703f8220dcffdeddb5e19dd032c766fbb8f/-/blob/libatomic/config/posix/lock.c#L91
https://sourcegraph.com/github.com/gcc-mirror/gcc@e1336703f8220dcffdeddb5e19dd032c766fbb8f/-/blob/libatomic/gstore.c#L100
I see  `__atomic_store` is implement by `pthread_mutex` + `memcpy`.

I perfer 
> or we can just leave with races and zero out non-atomically with memset, which is probably ok as well

than map, thanks!

Reply:
Fixed, thanks!